File: Shows/php/Stage/Proxy.php
<?php
include_once 'Subject.php';
include_once 'RealSubject.php';
class Proxy implements Subject
{
    private ?RealSubject $realSubject = null;
    public function request(): string
    {
        if ($this->realSubject === null) {
            $this->realSubject = new RealSubject();
        }
        return $this->realSubject->request();
    }
}

File: CSharp/Structural/Proxy/Example/Show/Show/Proxy.cs
using System;

namespace Show
{
    public class Proxy: ISubject
    {
        private RealSubject _realSubject;

        public Proxy(RealSubject realSubject)
        {
            _realSubject = realSubject;
        }

        public void Request()
        {
            if (CheckAccess())
            {
                _realSubject.Request();
                LogAccess();
            }
        }

        private bool CheckAccess()
        {
            // Here, you can put some real access control logic.
            Console.WriteLine("Proxy: Checking access before firing a real request.");
            return true;
        }

        private void LogAccess()
        {
            Console.WriteLine("Proxy: Logging the time of request.");
        }
    }
}